language/move-lang/src/naming/uses.rs

```diff
@@ -50,63 +61,105 @@ impl<'a> Context<'a> {
         }
     }
 
-    fn add_usage(&mut self, uses: &ModuleIdent, loc: Loc) {
-        if self.current_module.as_ref().unwrap() == uses || !self.modules.contains_key(uses) {
+    fn add_neighbor(&mut self, mident: ModuleIdent, dep_type: DepType, loc: Loc) {
+        let current_mident = self.current_module.clone().unwrap();
+        if current_mident == mident || !self.modules.contains_key(&mident) {
             return;
         }
+        let (node, new_neighbor) = match dep_type {
+            DepType::Use => (current_mident, mident),
+            DepType::Friend => (mident, current_mident),
+        };
 
         let m = self
             .neighbors
-            .entry(self.current_module.clone().unwrap())
+            .entry(node)
+            .or_insert_with(BTreeMap::new)
+            .entry(new_neighbor)
             .or_insert_with(BTreeMap::new);
-        if m.contains_key(uses) {
+        if m.contains_key(&dep_type) {
             return;
         }
+        m.insert(dep_type, loc);
+    }
 
-        m.insert(uses.clone(), loc);
+    fn add_usage(&mut self, mident: ModuleIdent, loc: Loc) {
+        self.add_neighbor(mident, DepType::Use, loc);
     }
 
-    fn dependency_graph(&self) -> DiGraphMap<&ModuleIdent, ()> {
-        let edges = self
-            .neighbors
-            .iter()
-            .flat_map(|(parent, children)| children.iter().map(move |(child, _)| (parent, child)));
-        DiGraphMap::from_edges(edges)
+    fn add_friend(&mut self, mident: ModuleIdent, loc: Loc) {
+        self.add_neighbor(mident, DepType::Friend, loc);
     }
 }
 
-fn cycle_error(context: &Context, cycle_ident: ModuleIdent) -> Error {
-    let cycle = shortest_cycle(&context.dependency_graph(), &cycle_ident);
+fn dependency_graph(
+    deps: &BTreeMap<ModuleIdent, BTreeMap<ModuleIdent, BTreeMap<DepType, Loc>>>,
+) -> DiGraphMap<&ModuleIdent, ()> {
+    let edges = deps
+        .iter()
+        .flat_map(|(parent, children)| children.iter().map(move |(child, _)| (parent, child)));
+    DiGraphMap::from_edges(edges)
+}
+
+fn cycle_error(
+    deps: &BTreeMap<ModuleIdent, BTreeMap<ModuleIdent, BTreeMap<DepType, Loc>>>,
+    cycle_ident: ModuleIdent,
+) -> Error {
+    let graph = dependency_graph(deps);
+    let cycle = shortest_cycle(&graph, &cycle_ident);
+
+    let mut cycle_strings: String = cycle
+        .windows(2)
+        .map(|pair| {
```

Comment:
Can you not do |(node, neighbor)| ?

Reply:
Yeah, that was my first trial as well but seemed not working because in `.windows(2)`, this 2 is actually a runtime thing and not statically known (i.e., we can replace it with a variable n and it will work too). So, it is unknown to Rust on how to unpack that.

Reply:
Ah I see, it gives a vec or something?

Reply:
Or some other random nonsense that implements indexing

Reply:
It gives a `Windows` that allows indexing